### PR TITLE
[11.x] Introduce method `Rule::array()`

### PR DESCRIPTION
**Pull request by Jacobs63**
This PR introduces a new `array` method to `\Illuminate\Validation\Rule`.

Currently, it is often cumbersome to validate multiple array keys through the `array` rule. At the same time, if intending to using constants, or enums, we may end up in defining our rule such as:
```php
['array:' . MyBackedEnum::VALUE->value . ',' . MyBackedEnum::VALUE_2->value]
```

The goal of this change is to improve code readability and prevent uncaught typos when validating array keys, but to also improve the existing functionalities by introducing support for `Arrayable` & lists of enums.

The newly introduced `ArrayRule` works very similar to the already existing `In` rule, by stringifying to the original `array` string rule, introducing no breaking changes.

In summary, the following is now possible:

```php
Rule::array();

Rule::array('key_1', 'key_2', 'key_3');

Rule::array(['key_1', 'key_2', 'key_3']);

Rule::array(collect(['key_1', 'key_2', 'key_3']));

Rule::array([UnitEnum::key_1, UnitEnum::key_2, UnitEnum::key_3]);

Rule::array([BackedEnum::key_1, BackedEnum::key_2, BackedEnum::key_3]);
```

### CODE REVIEW

**Comment by vaishnavyogesh** [comment]
@Jacobs63 why not use Rule::enum() for matching against an enum's cases?

**Comment by Jacobs63** [comment]
> @Jacobs63 why not use Rule::enum() for matching against an enum's cases?

The `array` rule does not serve for validating values, it can either validate, that:
1. The input value is of type array
2. The input value is of type array and only contains specific keys

Unlike `enum` rule, which validates valus and can validate that:
1. The input value matches the enum name (if UnitEnum)
2. The input value matches the enum value (if BackedEnum)

They're two different validation rules.

**Comment by vaishnavyogesh** [comment]
> > @Jacobs63 why not use Rule::enum() for matching against an enum's cases?
> 
> The `array` rule does not serve for validating values, it can either validate, that:
> 
> 1. The input value is of type array
> 2. The input value is of type array and only contains specific keys
> 
> Unlike `enum` rule, which validates values and can validate that:
> 
> 1. The input value matches the enum name (if UnitEnum)
> 2. The input value matches the enum value (if BackedEnum)
> 
> They're two different validation rules.

My bad! I misunderstood the usecase, thanks for the explaination though.

**Comment by osbre** [comment]
What's the difference between `Rule::in` and `Rule::array`?

**Comment by decadence** [comment]
> What's the difference between `Rule::in` and `Rule::array`? Why another method?

`array` defines keys that validated value (array) can have. `in` is list of allowed values for validated value.

**Comment by osbre** [comment]
@decadence Thank you. I wasn't aware there's an `array` in the first place 😅

https://laravel.com/docs/11.x/validation#rule-array
